jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala

```diff
@@ -56,26 +57,40 @@ private[jms] trait JmsConnector[S <: JmsSession] {
   override def preStart(): Unit = {
     // keep two elements since the time between initializing and connected can be very short.
     // always drops the old state, and keeps the most current (two) state(s) in the queue.
-    val pair = Source.queue[InternalConnectionState](2, OverflowStrategy.dropHead).preMaterialize()(this.materializer)
-    connectionStateQueue = pair._1
-    connectionStateSourcePromise.complete(Success(pair._2))
+    val (queue, source) =
+      Source
+        .queue[InternalConnectionState](2, OverflowStrategy.dropHead)
+        .toMat(BroadcastHub.sink(1))(Keep.both)
+        .run()(this.materializer)
+    connectionStateQueue = queue
+    connectionStateSourcePromise.complete(Success(source))
+
+    // add subscription to purge queued connection status events after the configured timeout.
+    val system: ActorSystem = ActorMaterializerHelper.downcast(materializer).system
+    after(jmsSettings.connectionStatusSubscriptionTimeout, system.scheduler) {
+      source.runWith(Sink.ignore)(this.materializer)
+    }
   }
 
   protected def finishStop(): Unit = {
     val update: InternalConnectionState => InternalConnectionState = {
       case JmsConnectorStopping(completion) => JmsConnectorStopped(completion)
-      case current => current
+      case stopped: JmsConnectorStopped => stopped
+      case current =>
+        JmsConnectorStopped(
+          Failure(new IllegalStateException(s"BUG: completing stage stop in unexpected state ${current.getClass}"))
+        )
     }
 
     closeSessions()
     val previous = updateStateWith(update)
-    connection(previous).foreach(_.close())
+    closeConnectionAsync(connection(previous))
```

Comment:
I wonder what the consequences of closing connections asynchronously without blocking could be.  I realize that that was what was done before, but ideally, we would chain any subsequent operations after asynchronously closing the connection object with `AsyncCallbacks`.  Maybe it's not very important.

Reply:
Let me think on it for a bit, I will come back to you on this.

Reply:
I had a look at the code and I think this is mainly relevant in [maybeReconnect](https://github.com/andreas-schroeder/alpakka/blob/jms-fix-connection-status-memory-leak/jms/src/main/scala/akka/stream/alpakka/jms/impl/JmsConnector.scala#L211). In other cases, the closing of the connection is within the stage teardown on successful completion or failure. In `maybeReconnect`, we could `scheduleOnce` in a `Future.andThen`. What do you think?

Reply:
Sounds good to me, but if you would like to work on it then it should be addressed in a new PR since your changes didn't introduce the issue in the first place.

Reply:
Alright. Then, let's leave it for the next PR.